File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> adj[204];
int visited[204];

void dfs(int here){
    visited[here] = 1;
    
    for(int there : adj[here]){
        if(visited[there]) continue;
        dfs(there);
    }
}

int solution(int n, vector<vector<int>> computers) {
    int ret = 0;
    
    for (int i=0; i<n; i++){
        for (int j=0; j<n; j++){
            if(i != j && computers[i][j]){
                adj[i].push_back(j);
            }
        }
    }
    
    for (int i=0; i<n; i++){
        if(!visited[i]){
            ret++;
            dfs(i);
        }
    }
    
    return ret;
}

File: 프로그래머스/3/43162. 네트워크/네트워크.js
const dfs = (here, adj, visited) => {
    visited.add(here)
    
    adj[here].forEach(there => {
        if(!visited.has(there)){
            dfs(there, adj, visited);
        }
    })
}

function solution(n, computers) {
    let ret = 0;
    
    const adj = Array.from({length: n}, () => []);
    const visited = new Set();
    
    for (let i=0; i<computers.length; i++){
        for (let j=0; j<computers[0].length; j++){
            if(i !== j && computers[i][j]){
                adj[i].push(j);
                adj[j].push(i);
            }
        }
    }
    
    for (let i=0; i<n; i++){
        if(!visited.has(i)){
            ret ++;
            dfs(i, adj, visited);
        }
    }
    
    return ret;
}